ext/include/opentelemetry/ext/zpages/threadsafe_span_data.h

```diff
@@ -0,0 +1,171 @@
+   */
+  opentelemetry::nostd::string_view GetName() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return name_;
+  }
+
+  /**
+   * Get the status for this span
+   * @return the status for this span
+   */
+  opentelemetry::trace::CanonicalCode GetStatus() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return status_code_;
+  }
+
+  /**
+   * Get the status description for this span
+   * @return the description of the the status of this span
+   */
+  opentelemetry::nostd::string_view GetDescription() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return status_desc_;
+  }
+
+  /**
+   * Get the start time for this span
+   * @return the start time for this span
+   */
+  opentelemetry::core::SystemTimestamp GetStartTime() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return start_time_;
+  }
+
+  /**
+   * Get the duration for this span
+   * @return the duration for this span
+   */
+  std::chrono::nanoseconds GetDuration() const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return duration_;
+  }
+
+  /**
+   * Get the attributes for this span
+   * @return the attributes for this span
+   */
+  const std::unordered_map<std::string, SpanDataAttributeValue> &GetAttributes()
+      const noexcept {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return attributes_;
+  }
+
+  void SetIds(opentelemetry::trace::TraceId trace_id,
+              opentelemetry::trace::SpanId span_id,
+              opentelemetry::trace::SpanId parent_span_id) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    trace_id_ = trace_id;
+    span_id_ = span_id;
+    parent_span_id_ = parent_span_id;
+  }
+
+  void SetAttribute(nostd::string_view key,
+                    const common::AttributeValue &value) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    attributes_[std::string(key)] = nostd::visit(converter_, value);
+  }
+
+  void AddEvent(nostd::string_view name,
+                core::SystemTimestamp timestamp) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    (void)name;
+    (void)timestamp;
+  }
+
+  void SetStatus(trace_api::CanonicalCode code,
+                 nostd::string_view description) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    status_code_ = code;
+    status_desc_ = std::string(description);
+  }
+
+  void SetName(nostd::string_view name) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    name_ = std::string(name);
+  }
+
+  void SetStartTime(
+      opentelemetry::core::SystemTimestamp start_time) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    start_time_ = start_time;
+  }
+
+  void SetDuration(std::chrono::nanoseconds duration) noexcept override {
+    std::unique_lock<std::mutex> lock(mutex_);
+    duration_ = duration;
+  }
+
+ private:
+  mutable std::mutex mutex_;
```

Comment:
I would use a [shared mutex](https://en.cppreference.com/w/cpp/thread/shared_mutex), so you can use shared locks.

Reply:
https://en.cppreference.com/w/cpp/thread/shared_mutex only in C++17, project runs C++11

Reply:
Oops, good catch!

Reply:
Also, wondering if the file should be under the sdk, since other features might want to use it if they're using span data between `OnStart` and `OnEnd`?

Reply:
For now it's only zpages if something else wants to use it, it can be moved